app/controllers/assignments_controller.rb

```diff
@@ -24,6 +24,7 @@ def edit
 
   # POST /assignments or /assignments.json
   def create
+    assignment_params[:repository_name].downcase!
```

Comment:
Maybe make this a `before_validation` callback, and maybe use this instead:
```
def generate_github_repo_name(input)
  repo_name = input.downcase
  
  # replace spaces/underscores with hyphens
  repo_name.gsub!(/[ _]/, '-')
  
  # remove any non-alphanumeric/hyphen chars
  repo_name.gsub!(/[^a-z0-9\-]/, '')
  
  # remove leading/trailing hyphens
  repo_name.gsub!(/^-+|-+$/, '')

  repo_name
end
```

Reply:
Thanks for the comment! Implemented this, by providing a `before_validation` hook in the Assignment model.

However, I had to make a small adjustment in the controller. In the `AssignmentController#create` the `assignment_repo_init` function was called before the `@assignment.save` statement (where validation hooks are called).

As a result, the validation hook was called after the repo was created in our organization, which was causing a few errors for the cases of incorrect repository name entries (for example, those that have whitespaces, etc. ). 

So, I shifted the `assignment_repo_init` call slightly after the `@assignment.save' statement. This allows the repo to be created after the validation hooks are called. (See below)

```
  def create
    @assignment = Assignment.new(assignment_params)
    github_token = session[:github_token]

    respond_to do |format|
      if @assignment.save
        @assignment.assignment_repo_init(github_token)
        format.html { redirect_to @assignment, notice: "Assignment was successfully created." }
        format.json { render :show, status: :created, location: @assignment }
      else
        format.html { render :new, status: :unprocessable_entity }
        format.json { render json: @assignment.errors, status: :unprocessable_entity }
      end
    end
  end